### PR TITLE
Add support for responding locally (e.g with an error).

### PR DESCRIPTION
**Pull request by jplevyak**
Added support for clearing the route cache.
Converted GrpcStatus to use the envoy names.

Signed-off-by: John Plevyak <jplevyak@gmail.com>

For an explanation of how to fill out the fields, please see the relevant section
in [PULL_REQUESTS.md](https://github.com/envoyproxy/envoy/blob/master/PULL_REQUESTS.md)

Description:
Risk Level:
Testing:
Docs Changes:
Release Notes:
[Optional Fixes #Issue]
[Optional Deprecated:]

### CODE REVIEW

**Comment by jplevyak** [comment]
https://github.com/envoyproxy/envoy-wasm/issues/74

**Comment by jplevyak** [comment]
I rebuilt the .wasm files with the latest emscripten, and the tests worked for me locally.  It seems like there is some issue with the magic numbers in the module.

File: source/extensions/common/wasm/null/wasm_api_impl.h [hunk 1]
```diff
@@ -73,6 +73,17 @@ inline void proxy_setMetadataStruct(MetadataType type, const char* name_ptr, siz
 // Continue
 inline void proxy_continueRequest() { continueRequestHandler(current_context_); }
 inline void proxy_continueResponse() { continueResponseHandler(current_context_); }
+inline void proxy_sendLocalReply(uint32_t response_code, const char* response_code_details_ptr,
```
Comment (PiotrSikora) [inline source/extensions/common/wasm/null/wasm_api_impl.h]: Nit: I'd prefer if this was called `sendLocalResponse`.

File: test/extensions/wasm/wasm_test.cc [hunk 2]
```diff
@@ -175,9 +175,9 @@ TEST_P(WasmTest, EmscriptenVersion) {
   uint32_t major = 9, minor = 9, abi_major = 9, abi_minor = 9;
   EXPECT_TRUE(wasm->getEmscriptenVersion(&major, &minor, &abi_major, &abi_minor));
   EXPECT_EQ(major, 0);
-  EXPECT_EQ(minor, 0);
+  EXPECT_TRUE(minor >= 0 && minor <= 1);
   EXPECT_EQ(abi_major, 0);
-  EXPECT_EQ(abi_minor, 1);
+  EXPECT_TRUE(abi_minor >= 0 && abi_minor <= 3);
```
Comment (PiotrSikora) [inline test/extensions/wasm/wasm_test.cc]: Unrelated changes?

File: source/extensions/common/wasm/null/wasm_api_impl.h (hunk 1 above)
Comment (jplevyak) [reply source/extensions/common/wasm/null/wasm_api_impl.h]: I did this to match the envoy name, but I agree, that is a better name.

File: test/extensions/wasm/wasm_test.cc (hunk 2 above)
Comment (jplevyak) [reply test/extensions/wasm/wasm_test.cc]: The provides support for up to 1.38.39 of emscripten.  Yes, it is unrelated.  I can pull it out if you want.

**Comment by jplevyak** [comment]
The asan test fails with a timeout and

tcpdump: No such file or directory

File: source/extensions/common/wasm/null/wasm_api_impl.h (hunk 1 above)
Comment (PiotrSikora) [reply source/extensions/common/wasm/null/wasm_api_impl.h]: We don't match Envoy names for most part (e.g. request/response vs encode/decode), so I don't see a strong need to do it here either.

File: test/extensions/wasm/wasm_test.cc (hunk 2 above)
Comment (PiotrSikora) [reply test/extensions/wasm/wasm_test.cc]: I'd prefer to have separate PRs for working features that clearly describe intention behind the change, than sneaked in changes and/or code dumps.

**Comment by PiotrSikora** [comment]
Yeah, asan has been timing out from time to time lately.

**Comment by PiotrSikora** [comment]
Can you merge master? Thanks!

File: source/extensions/common/wasm/null/wasm_api_impl.h (hunk 1 above)
Comment (jplevyak) [reply source/extensions/common/wasm/null/wasm_api_impl.h]: Changed.  I almost just called it "respond" but that is a bit terse.

File: test/extensions/wasm/wasm_test.cc (hunk 2 above)
Comment (jplevyak) [reply test/extensions/wasm/wasm_test.cc]: OK, I'll pull that out.

Comment (jplevyak) [reply test/extensions/wasm/wasm_test.cc]: https://github.com/envoyproxy/envoy-wasm/pull/83

**Comment by jplevyak** [comment]
updated to master

File: api/wasm/cpp/proxy_wasm_intrinsics.h [hunk 3]
```diff
@@ -10,7 +10,7 @@
 #endif
 
 #include <cstdint>
-using EnumType = uint32_t;
+using EnumType = int32_t;
```
Comment (silentdai) [inline api/wasm/cpp/proxy_wasm_intrinsics.h]: Why are you making this change? For the Invalid as "-1"? You could have replace it by "std::numeric_limits<EnumType>::max()"

File: api/wasm/cpp/proxy_wasm_impl.h [hunk 4]
```diff
@@ -631,9 +631,19 @@ inline google::protobuf::Struct Context::getResponseMetadataStruct(StringView na
   return getMetadataStruct(MetadataType::Response, name);
 }
 
-// Continue
+// Continue/Respond/Route
 inline void continueRequest() { proxy_continueRequest(); }
 inline void continueResponse() { proxy_continueResponse(); }
+inline void sendLocalResponse(uint32_t response_code, StringView response_code_details, StringView body,
+    const HeaderStringPairs& additional_response_headers, uint32_t grpc_status = GrpcStatus::InvalidCode) {
```
Comment (silentdai) [inline api/wasm/cpp/proxy_wasm_impl.h]: there is an implicit int32_t to uint32_t. See my suggestion in EnumType

File: source/extensions/common/wasm/wasm.cc [hunk 5]
```diff
@@ -431,6 +431,36 @@ void continueResponseHandler(void* raw_context) {
...
+                              Word response_code_details_size, Word body_ptr, Word body_size,
+                              Word additional_response_header_pairs_ptr,
+                              Word additional_response_header_pairs_size, Word grpc_code) {
+  auto context = WASM_CONTEXT(raw_context);
+  auto details =
+      context->wasmVm()->getMemory(response_code_details_ptr, response_code_details_size);
+  auto body = context->wasmVm()->getMemory(body_ptr, body_size);
+  auto additional_headers = toPairs(context->wasmVm()->getMemory(
+      additional_response_header_pairs_ptr, additional_response_header_pairs_size));
+  std::function<void(Http::HeaderMap & headers)> modify_headers =
```
Comment (silentdai) [inline source/extensions/common/wasm/wasm.cc]: is this formatted?

Comment (silentdai) [inline source/extensions/common/wasm/wasm.cc]: use auto to obtain lambda type. lambda will convert to function when it is needed

File: source/extensions/common/wasm/wasm.cc [hunk 6]
```diff
@@ -1650,6 +1684,7 @@ void Wasm::registerCallbacks() {
     _REGISTER(___cxa_throw);
     _REGISTER(___cxa_pure_virtual);
     _REGISTER(___cxa_allocate_exception);
+    _REGISTER(___cxa_uncaught_exception);
```
Comment (silentdai) [inline source/extensions/common/wasm/wasm.cc]: For my curiosity: could you explain why we need it here? Link is also good. Thanks!

File: source/extensions/common/wasm/null/null.h [hunk 7]
```diff
@@ -15,7 +15,7 @@ using FilterHeadersStatus = Http::FilterHeadersStatus;
 using FilterMetadataStatus = Http::FilterMetadataStatus;
 using FilterTrailersStatus = Http::FilterTrailersStatus;
 using FilterDataStatus = Http::FilterDataStatus;
-using GrpcStatus = ProtobufUtil::error::Code;
+using GrpcStatus = Envoy::Grpc::Status::GrpcStatus;
```
Comment (silentdai) [inline source/extensions/common/wasm/null/null.h]: Naive question: why?

File: api/wasm/cpp/proxy_wasm_intrinsics.h (hunk 3 above)
Comment (jplevyak) [reply api/wasm/cpp/proxy_wasm_intrinsics.h]: I am matching the envoy GrpcStatus class so we can use that one when using the null VM.

File: source/extensions/common/wasm/null/null.h (hunk 7 above)
Comment (jplevyak) [reply source/extensions/common/wasm/null/null.h]: Because the latter defines an invalid status which I can use to indicate that the gRPC status is not provided because the envoy function takes std::optional<GrpcStatus> but the ABI is based on raw words, so std::optional is not available and I need a sentinel value.

File: source/extensions/common/wasm/wasm.cc (hunk 6 above)
Comment (jplevyak) [reply source/extensions/common/wasm/wasm.cc]: This is to support emscripten 1.38.39.  This will disappear from this PR when #83 is merged.

**Comment by jplevyak** [comment]
I am going to hold off merging till #83 is merged.

File: api/wasm/cpp/proxy_wasm_enums.h [hunk 8]
```diff
@@ -34,24 +34,25 @@ enum class HeaderMapType : EnumType {
   GrpcReceiveTrailingMetadata = 6,  // Immutable
 };
 enum GrpcStatus : EnumType {
```
[the remark this answers is not in the document]
Comment (jplevyak) [reply api/wasm/cpp/proxy_wasm_enums.h]: done

File: api/wasm/cpp/proxy_wasm_impl.h (hunk 4 above)
Comment (jplevyak) [reply api/wasm/cpp/proxy_wasm_impl.h]: Made the conversion explicit.

File: source/extensions/common/wasm/wasm.cc (hunk 5 above)
Comment (jplevyak) [reply source/extensions/common/wasm/wasm.cc]: yes, but now it is changed as per below.

Comment (jplevyak) [reply source/extensions/common/wasm/wasm.cc]: done

**Comment by silentdai** [comment]
/lgtm

Thanks!
